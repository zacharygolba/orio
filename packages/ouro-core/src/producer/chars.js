// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'
import type { Producer } from 'ouro-types'

@ToString
@AsIterator
export default class Chars implements Producer<*> {
  /*:: @@iterator: () => Iterator<string> */
  cursor: number
  source: string

  constructor(source: string) {
    this.cursor = 0
    this.source = source
  }

  drop(): void {
    this.cursor = 0
    this.source = ''
  }

  next(): IteratorResult<string, void> {
    const { cursor, source } = this

    if (cursor >= source.length) {
      return result.done()
    }

    this.cursor += 1
    return result.next(source.substr(cursor, 1))
  }
}
