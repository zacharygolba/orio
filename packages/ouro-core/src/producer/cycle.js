// @flow

import * as result from '../result'

import ProducerBase from './base'
import type { IndexedCollection } from './'

export default class CycleProducer<T> extends ProducerBase<T> {
  source: IndexedCollection<T>
  state: number

  constructor(source: IndexedCollection<T>) {
    super()
    this.source = source
    this.state = 0
  }

  next(): IteratorResult<T, void> {
    if (this.state >= this.source.length) {
      this.state = 0
    }

    const state = this.state
    this.state += 1

    return result.next(this.source[state])
  }
}
