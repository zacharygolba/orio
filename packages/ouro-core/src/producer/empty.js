// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'
import type { Producer } from 'ouro-types'

@ToString
@AsIterator
export default class Empty implements Producer<*> {
  /*:: @@iterator: () => Iterator<*> */
  value: void

  drop(): void {
    this.value = undefined
  }

  next(): IteratorResult<*, void> {
    return result.done(this.value)
  }
}
