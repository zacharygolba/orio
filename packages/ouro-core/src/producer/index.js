// @flow

import Indexed from './indexed'
import Unbound from './unbound'

export * from './range'
export { default as Cycle } from './cycle'
export { default as Indexed } from './indexed'
export { default as Repeat } from './repeat'
export { default as Unbound } from './unbound'
export type { IndexedCollection } from './indexed'

export function createProducer<T>(source: any): Iterator<T> {
  if (source == null) {
    return new Indexed([])
  }

  if (Array.isArray(source) || typeof source === 'string') {
    return new Indexed(source)
  }

  if (source instanceof Set) {
    return source.values()
  }

  if (source instanceof Map) {
    return source.entries()
  }

  if (typeof source[Symbol.iterator] === 'function') {
    return new Unbound(source)
  }

  return new Indexed([source])
}
