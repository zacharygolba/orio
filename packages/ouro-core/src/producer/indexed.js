// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'

import type { Drop, IndexedCollection } from '../types'

@ToString
@AsIterator
export default class Indexed<T> implements Drop, Iterator<T> {
  /*:: @@iterator: () => Iterator<T> */
  index: number
  source: IndexedCollection<T>

  constructor(source: IndexedCollection<T>) {
    this.index = 0
    this.source = source
  }

  drop(): void {
    this.index = 0
    this.source = []
  }

  next(): IteratorResult<T, void> {
    const { index, source } = this

    if (index >= source.length) {
      return result.done()
    }

    this.index += 1
    return result.next(source[index])
  }
}
