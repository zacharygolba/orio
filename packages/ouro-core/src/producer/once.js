// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'
import type { Producer } from 'ouro-types'

@ToString
@AsIterator
export default class Once<T> implements Producer<T> {
  /*:: @@iterator: () => Iterator<T> */
  done: boolean
  value: T

  constructor(value: T) {
    this.done = false
    this.value = value
  }

  drop(): void {
    this.done = true
  }

  next(): IteratorResult<T, void> {
    if (this.done) {
      return result.done()
    }

    this.done = true
    return result.next(this.value)
  }
}
