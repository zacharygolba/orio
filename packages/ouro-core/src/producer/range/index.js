// @flow

export { default as Chars } from './chars'
export { default as Numbers } from './numbers'
