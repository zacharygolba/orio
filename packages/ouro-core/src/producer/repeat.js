// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'

@ToString
@AsIterator
export default class Repeat<T> implements Iterator<T> {
  /*:: @@iterator: () => Iterator<T> */
  value: T

  constructor(value: T) {
    this.value = value
  }

  next(): IteratorResult<T, void> {
    return result.next(this.value)
  }
}
