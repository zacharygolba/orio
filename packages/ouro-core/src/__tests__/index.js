// @flow

import { Chars, Cycle, Numbers } from '../producer'
import Ouro from '../ouro'
import * as ouro from '../'

test('.chars()', () => {
  expect(ouro.chars()).toBeInstanceOf(Ouro)
  expect(ouro.chars('a')).toBeInstanceOf(Ouro)
  expect(ouro.chars('a', 'z')).toBeInstanceOf(Ouro)

  {
    const iter = ouro.chars('a', 'z')

    expect(iter.producer).toBeInstanceOf(Chars)
  }
})

test('.cycle()', () => {
  const iter = ouro.cycle([1, 2, 3])

  expect(iter).toBeInstanceOf(Ouro)
  expect(iter.producer).toBeInstanceOf(Cycle)
})

test('.from()', () => {
  expect(ouro.from()).toBeInstanceOf(Ouro)
  expect(ouro.from([1, 2, 3])).toBeInstanceOf(Ouro)
})

test('.of()', () => {
  expect(ouro.of()).toBeInstanceOf(Ouro)

  {
    const iter = ouro.of(1, 2, 3)

    expect(iter).toBeInstanceOf(Ouro)
    expect(iter.producer).toMatchSnapshot()
  }
})

test('.range()', () => {
  expect(ouro.range()).toBeInstanceOf(Ouro)
  expect(ouro.range(1)).toBeInstanceOf(Ouro)
  expect(ouro.range(1, 2)).toBeInstanceOf(Ouro)

  {
    const iter = ouro.range(1, 2)

    expect(iter.producer).toBeInstanceOf(Numbers)
  }
})

test('.repeat()', () => {
  expect(ouro.repeat()).toBeInstanceOf(Ouro)
})
