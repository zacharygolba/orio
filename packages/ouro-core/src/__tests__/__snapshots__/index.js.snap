// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.of() 1`] = `
Indexed {
  "index": 0,
  "source": Array [
    1,
    2,
    3,
  ],
}
`;
